### PR TITLE
Add self.logger to python block gateway

### LINKED ISSUE #6284: Subpar logger integration in Python block gateway >= 3.10
### What happened?

In the process of rewriting the logging, I did not take care to forward the `d_logger` member of the C++ block to python. 

That leads to ugly constraints: users have to create their own logger and use that.

### System Information

main

### GNU Radio Version

3.11-git (main)

### Specific Version

3.11xxxxgit

### Steps to Reproduce the Problem

make python block

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by marcusmueller**
- logging: add in-memory logger with filtering
- runtime: Add .logger property to block gateway
- logger: removed misleading comment

## Description

The Python gateway was missing the capability to offer the `d_logger` of any GR block to pythonland.

This fixes it.

## Related Issue

Closes #6284

## Which blocks/areas does this affect?

Python blocks

## Testing Done
To verify operability, wrote a little plaything in-memory logging backend, from which logged messages can be retrieved.

## Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md).
- [x] I have squashed my commits to have one significant change per commit. 
- [x] I [have signed my commits before making this PR](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md#dco-signed)
- [x] My code follows the code style of this project. See [GREP1.md](https://github.com/gnuradio/greps/blob/main/grep-0001-coding-guidelines.md).
- [ ] I have updated [the documentation](https://wiki.gnuradio.org/index.php/Main_Page#Documentation) where necessary.
- [x] I have added tests to cover my changes, and all previous tests pass.

Documentation needs discussion: The wiki logging page should sure taut it, but the question is for which versions.

### CODE REVIEW

**Comment by marcusmueller** [comment]
Re: backport to 3.10: Adds new functionality, but should not break ABI.

**Comment by marcusmueller** [comment]
hm tests pass locally, both natively and in VM. huh. stressing CPUs doesn't help reproduce the failures either. And OSX worries me.

**Comment by marcusmueller** [comment]
@ryanvolz I'm seeing Segfaults on OSX that I can't explain (nor explain away). Is there some easy way to get a backtrace or something similar of a selected build?

**Comment by ryanvolz** [comment]
> @ryanvolz I'm seeing Segfaults on OSX that I can't explain (nor explain away). Is there some easy way to get a backtrace or something similar of a selected build?

No easy way that I know of, unless a file would have already been produced in the build directory that could be viewed in the downloaded artifact. Other things that come to mind include modifying `build.sh` to do something useful for debugging or tweaking the Action to [provide ssh access](https://github.com/marketplace/actions/debugging-with-ssh) .

**Comment by ryanvolz** [comment]
But the segfault is probably this: https://github.com/gnuradio/gnuradio/issues/6283

**Comment by marcusmueller** [comment]
Oh! Am I writing the first unit test for a Python block?!?!? Oh dear...

**Comment by ryanvolz** [comment]
No, but the few others are disabled 😬.

**Comment by willcode** [comment]
What is the effect of changing the defaults from "off" to "info"? Some blocks (quick search: compass.py, digitalnumbercontrol.py) would need some quieting down.

**Comment by marcusmueller** [comment]
> What is the effect of changing the defaults from "off" to "info"? Some blocks (quick search: compass.py, digitalnumbercontrol.py) would need some quieting down.

I did that because I realized that "off" is probably not the default users would want. And it's probably not in our interest to have that be the default, either, as it would make it harder to get diagnostics from users (who we then would need to instruct how to change config files).

I hadn't thought about checking for "noisy" blocks; the question is whether we would want to fix them by reducing the log level of their frequent output to `trace`?

**Comment by willcode** [comment]
Changing noise blocks to `trace` sounds good.

On `[[nodiscard]]` in general, I see the benefit of the warning, but if we start adding decorators everywhere the code is going to get harder to read. The recommendation for use on `singleton()` ignores the fact that some long-running initializations (not the case here) might want to be run before the singleton is used for anything.

**Comment by marcusmueller** [comment]
> The recommendation for use on singleton() ignores the fact that some long-running initializations (not the case here) might want to be run before the singleton is used for anything.

oh, fair point. Also, if it's already been done before, using it as discarded value has zero cost, so will remove.

File: gr-digital/lib/crc_check_impl.h [hunk 1]
```diff
@@ -44,7 +44,7 @@ class crc_check_impl : public crc_check
                      gr_vector_int& ninput_items,
                      gr_vector_const_void_star& input_items,
                      gr_vector_void_star& output_items);
-    void msg_handler(pmt::pmt_t pmt_msg);
+    void msg_handler(const pmt::pmt_t& pmt_msg);
```
Comment (willcode) [inline gr-digital/lib/crc_check_impl.h:47]: API backward compatible?

**Review by willcode — APPROVED** [review]
As long as the `const` is backward compatible.

I didn't see the `debug_click` guards before. Your changes make more sense than that anyway. Sorry that triggered all the reformatting.

File: gr-digital/lib/crc_check_impl.h (hunk 1 above)
Comment (marcusmueller) [reply gr-digital/lib/crc_check_impl.h:47]: nope, hence singled-out in discardable commit.

Comment (willcode) [reply gr-digital/lib/crc_check_impl.h:47]: Is the `const` required for the `ok_port`/`fail_port` initializaiton, or just a good idea?

Comment (marcusmueller) [reply gr-digital/lib/crc_check_impl.h:47]: ah, no, it's not required, but I didn't want to make two miniature performance-enhancing commits, can split if you actually want to backport only the static initialization enhancement.

**Review by mormj — APPROVED** [review]

**Comment by mormj** [comment]
This is consistently failing the python_logging test for conda macOS:
```
 The following tests FAILED:
	 17 - qa_python_logging (Failed)
Errors while running CTest
Traceback (most recent call last):
  File "/Users/runner/miniforge3/bin/conda-mambabuild", line 10, in <module>
    sys.exit(main())
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/boa/cli/mambabuild.py", line 239, in main
    call_conda_build(action, config)
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/boa/cli/mambabuild.py", line 211, in call_conda_build
    result = api.build(
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda_build/api.py", line 186, in build
    return build_tree(
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda_build/build.py", line 3091, in build_tree
    packages_from_this = build(metadata, stats,
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda_build/build.py", line 2214, in build
    utils.check_call_env(cmd, env=env, rewrite_stdout_env=rewrite_env,
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda_build/utils.py", line 412, in check_call_env
    return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda_build/utils.py", line 392, in _func_defaulting_env_to_os_environ
    raise subprocess.CalledProcessError(proc.returncode, _args)
subprocess.CalledProcessError: Command '['/bin/bash', '-o', 'errexit', '/Users/runner/miniforge3/conda-bld/gnuradio-dev_1667852926673/work/conda_build.sh']' returned non-zero exit status 8.
Error: Process completed with exit code 1.
```

Not sure what would be unique about macOS here

**Comment by ryanvolz** [comment]
> This is consistently failing the python_logging test for conda macOS:

Should be fixed by #6352, but one or the other will have to go first and then rebase the remaining PR.

**Comment by mormj** [comment]
Closing this one, but merging the rebased version over at #6364

**Pull request closed without merging**